src/probe/probe.c

```diff
@@ -15,42 +22,529 @@
+	uint32_t j;
+	uint32_t buffer_id;
+	uint32_t first_free;
+	uint32_t dma_found;
+	struct ipc_comp_dev *dev;
+
+	tracev_probe("probe_point_add() count = %u", count);
+
+	if (!_probe) {
+		trace_probe_error("probe_point_add() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	for (i = 0; i < count; i++) {
+		tracev_probe("\tprobe[%u] buffer_id = %u, purpose = %u, stream_tag = %u",
+			     i, probe[i].buffer_id, probe[i].purpose,
+			     probe[i].stream_tag);
+
+		if (probe[i].purpose == PROBE_PURPOSE_EXTRACTION &&
+		    _probe->ext_dma.stream_tag == PROBE_DMA_INVALID) {
+			trace_probe_error("probe_point_add() error: Setting probe for extraction, while extraction DMA not enabled.");
+
+			return -EINVAL;
+		}
+
+		dev = ipc_glb_get_comp_by_id(probe[i].buffer_id);
+		if (!dev) {
+			trace_probe_error("probe_point_add() error: No device with ID %u found.",
+					  probe[i].buffer_id);
+
+			return -EINVAL;
+		}
+
+		if (dev->type != COMP_TYPE_BUFFER) {
+			trace_probe_error("probe_point_add() error: Device ID %u is not a buffer.",
+					  probe[i].buffer_id);
+
+			return -EINVAL;
+		}
+
+		first_free = CONFIG_PROBE_POINTS_MAX;
+
+		for (j = 0; j < CONFIG_PROBE_POINTS_MAX; j++) {
+			if (_probe->probe_points[j].stream_tag ==
+			    PROBE_POINT_INVALID) {
+				if (first_free == CONFIG_PROBE_POINTS_MAX)
+					first_free = j;
+
+				continue;
+			}
+
+			buffer_id = _probe->probe_points[j].buffer_id;
+			if (buffer_id == probe[i].buffer_id) {
+				if (_probe->probe_points[j].purpose ==
+				    probe[i].purpose) {
+					trace_probe_error("probe_point_add() error: Probe already attached to buffer %u with purpose %u",
+							  buffer_id,
+							  probe[i].purpose);
+
+					return -EINVAL;
+				}
+			}
+		}
+
+		if (first_free == CONFIG_PROBE_POINTS_MAX) {
+			trace_probe_error("probe_point_add() error: Maximum number of probe points connected aleady: " META_QUOTE(CONFIG_PROBE_POINTS_MAX));
+
+			return -EINVAL;
+		}
+
+		/* if connecting injection probe, check for associated DMA */
+		if (probe[i].purpose == PROBE_PURPOSE_INJECTION) {
+			dma_found = 0;
+
+			for (j = 0; j < CONFIG_PROBE_DMA_MAX; j++) {
+				if (_probe->inject_dma[j].stream_tag !=
+				    PROBE_DMA_INVALID &&
+				    _probe->inject_dma[j].stream_tag ==
+				    probe[i].stream_tag) {
+					dma_found = 1;
+					break;
+				}
+			}
+
+			if (!dma_found) {
+				trace_probe_error("probe_point_add() error: No DMA with stream tag %u found for injection.",
+						  probe[i].stream_tag);
+
+				return -EINVAL;
+			}
+		}
+
+		_probe->probe_points[first_free].buffer_id = probe[i].buffer_id;
+		_probe->probe_points[first_free].purpose = probe[i].purpose;
+		_probe->probe_points[first_free].stream_tag =
+			probe[i].stream_tag;
+
+		/* TODO: Hook up callbacks to buffer (dev->cb) and DMA */
+	}
+
 	return 0;
 }
 
 int probe_point_info(struct sof_ipc_probe_info_params *data, uint32_t max_size)
 {
-	return 0;
+	uint32_t i = 0;
+	uint32_t j = 0;
+	uint32_t *size = &data->rhdr.hdr.size;
```

Comment:
It's better not to use ```size``` variable here and above as it hides the true intention of the code, would be far easier to use ```rhd->.hdr.size``` directly, that way readers will know we are referring to the reply header size.

Reply:
done